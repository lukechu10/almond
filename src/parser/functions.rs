//! Parsing for JS functions and programs.

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom::InputLength;
use nom_locate::position;

/// Parses a complete JS program
pub fn parse_program(s: Span) -> ParseResult<Node> {
    let (s, start) = position(s)?;
    let (s, _) = sp0(s)?; // eat all preceding whitespace

    let (s, body) = parse_function_body_inner(s)?;
    if s.input_len() != 0 {
        panic!("The source code was not completely parsed. This is a bug with the parser.\nUnparsed code starting on line {}", s.location_line());
    }

    let (s, end) = position(s)?; // Program loc should include all trailing whitespace
    Ok((s, NodeKind::Program { body }.with_pos(start, end)))
}

pub fn parse_declaration(s: Span) -> ParseResult<Node> {
    parse_function_declaration(s)
}

pub fn parse_function_declaration(s: Span) -> ParseResult<Node> {
    let parse_function_declaration_signature = pair(
        separated_pair(
            ws0(opt(keyword_async)),
            ws1(keyword_function),
            parse_identifier,
        ),
        delimited(ws0(tag("(")), parse_formal_param_list, ws0(tag(")"))),
    );

    context(
        "function declaration",
        map(
            spanned(pair(
                parse_function_declaration_signature,
                parse_function_body,
            )),
            |((((is_async, id), params), body), start, end)| {
                NodeKind::FunctionDeclaration {
                    function: Function {
                        id: Box::new(Some(id)),
                        params,
                        body: Box::new(body),
                        is_async: is_async.is_some(),
                    },
                }
                .with_pos(start, end)
            },
        ),
    )(s)
}

pub fn parse_function_expr(s: Span) -> ParseResult<Node> {
    let parse_function_expr_signature = pair(
        separated_pair(
            ws0(opt(keyword_async)),
            ws0(keyword_function),
            opt(parse_identifier),
        ),
        delimited(ws0(tag("(")), parse_formal_param_list, ws0(tag(")"))),
    );

    context(
        "function expression",
        map(
            spanned(pair(parse_function_expr_signature, parse_function_body)),
            |((((is_async, id), params), body), start, end)| {
                NodeKind::FunctionExpression {
                    function: Function {
                        id: Box::new(id),
                        params,
                        body: Box::new(body),
                        is_async: is_async.is_some(),
                    },
                }
                .with_pos(start, end)
            },
        ),
    )(s)
}

pub fn parse_function_body(s: Span) -> ParseResult<Node> {
    context(
        "function body",
        map(
            spanned(delimited(
                ws0(tag("{")),
                parse_function_body_inner,
                ws0(tag("}")),
            )),
            |(body, start, end)| NodeKind::BlockStatement { body }.with_pos(start, end),
        ),
    )(s)
}

pub fn parse_function_body_inner(s: Span) -> ParseResult<Vec<Node>> {
    let parse_directive_list = many0(parse_directive);
    let parse_source_element_list = many0(parse_source_elem);

    context(
        "function body inner",
        map(
            pair(parse_directive_list, parse_source_element_list),
            |(mut directives, stmts)| {
                directives.extend(stmts);
                directives
            },
        ),
    )(s)
}

pub fn parse_formal_param_list(s: Span) -> ParseResult<Vec<Node>> {
    terminated(
        separated_list0(ws0(char(',')), parse_formal_param),
        // trailing comma
        opt(ws0(char(','))),
    )(s)
}

pub fn parse_formal_param(s: Span) -> ParseResult<Node> {
    parse_identifier(s)
}

/// Example: `"use strict";`
pub fn parse_directive(s: Span) -> ParseResult<Node> {
    let parse_directive_literal = spanned(terminated(ws0(string_lit), opt(ws0(semi))));
    map(parse_directive_literal, |(node, start, end)| {
        match &node.kind {
            NodeKind::Literal { value } => match value {
                LiteralValue::String(directive) => {
                    let directive = directive.clone();
                    NodeKind::ExpressionStatement {
                        expression: Box::new(node),
                        directive: Some(directive),
                    }
                    .with_pos(start, end)
                }
                _ => unreachable!(
                    "string_lit should return a literal with value LiteralValue::String"
                ),
            },
            _ => unreachable!("string_lit should return a Node::Identifier"),
        }
    })(s)
}

pub fn parse_source_elem(s: Span) -> ParseResult<Node> {
    alt((parse_declaration, parse_stmt))(s)
}

#[cfg(test)]
mod tests {
    use super::*;
    use insta::assert_json_snapshot;

    #[test]
    fn test_directive() {
        assert_json_snapshot!(parse_directive("\"use strict\"".into()).unwrap().1);
    }

    #[test]
    fn test_function_declaration() {
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    "use strict"
                    return 1
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a() {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a(param1, param2) {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(r#"function a(param1, param2,) {}"#.into())
                .unwrap()
                .1
        );
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    "use strict"
                    "also directive"
                    a;
                    "not directive"
                    return 1
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function fib() {
                    if (x === 0) return 0;
                    else if (x === 1) return 1;
                    else return fib(x - 1) + fib(x - 2);
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"function foo() {
                    // stuff
                }; foo();"# // note the `;` after the function declaration. Should be parsed as an `EmptyStatement`.
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_nested_function_declaration() {
        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    function b() {
                        a();
                    }
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_function_expr() {
        parse_function_declaration(
            r#"function a() {
                // function expressions should be wrapped up in parenthesis
                function () {
                    a();
                }
            }"#
            .into(),
        )
        .unwrap_err();

        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    // anonymous function:
                    (function () {
                        a();
                    })
                }"#
                .into()
            )
            .unwrap()
            .1
        );

        assert_json_snapshot!(
            parse_function_declaration(
                r#"function a() {
                    // function expression
                    0, function b() {}
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn parse_immediately_invoked_function_expr() {
        assert_json_snapshot!(
            parse_stmt(
                r#"(function () {
                    123; // something
                })()"#
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_program() {
        assert_json_snapshot!(
            parse_program(
                r#"var x = 1;
                function foo() {
                    return x;
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"var x = 1;
                function foo() {
                    if (x) {
                        return x;
                    }
                }"#
                .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"
                function fib() {
                    if (x === 0) return 0;
                    else if (x === 1) return 1;
                    else return fib(x - 1) + fib(x - 2);
                }
                var x = fib(10);"#
                    .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"function foo() {
                    test;
                    // abc "123"-
                    // test
                    foo();
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_empty_function() {
        assert_json_snapshot!(
            parse_program(
                r#"function foo(callback) {
                    // A👍a
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn parse_program_immediately_invoked_function_expr() {
        assert_json_snapshot!(
            parse_program(
                r#"(function () {
                    123; // something
                })()"#
                    .into()
            )
            .unwrap()
            .1
        );
        assert_json_snapshot!(
            parse_program(
                r#"(function (global, factory) {
                    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
                        typeof define === 'function' && define.amd ? define(['exports'], factory) :
                            (global = global || self, factory(global.Library = {}));
                }(this, (function (exports) {})));"#
                    .into()
            )
            .unwrap()
            .1
        );
    }

    #[test]
    fn test_async_function() {
        assert_json_snapshot!(
            parse_program(
                r#"async function foo(callback) {
                    await test();
                }"#
                .into()
            )
            .unwrap()
            .1
        );
    }
}
