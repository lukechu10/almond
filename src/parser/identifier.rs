//! Parsing for JS identifiers

use crate::ast::*;
use crate::parser::util::*;
use crate::parser::*;
use nom_locate::position;

/// Parses any 4 hex digits unicode escape sequence.
pub fn unicode_esc_seq(s: Span) -> ParseResult<char> {
    let (s, hex_str) = recognize(count(one_of("1234567890abcdefABCDEF"), 4))(s)?;
    let hex_u32 = u32::from_str_radix(*hex_str, 16).unwrap(); // FIXME
    let char = std::char::from_u32(hex_u32).unwrap(); // FIXME
    Ok((s, char))
}

/// Parses `"\\u"` followed by any 4 hex digits unicode escape sequence.
pub fn identifier_unicode_esc_sequence(s: Span) -> ParseResult<char> {
    preceded(tag("\\u"), unicode_esc_seq)(s)
}

pub fn identifier_start(s: Span) -> ParseResult<char> {
    verify(alt((identifier_unicode_esc_sequence, anychar)), |c: &char| match c {
        // These are already matched by is_xid_start but are explicitly at the beginning for performance reasons.
        'a'..='z' => true,
        'A'..='Z' => true,
        c if unicode_xid::UnicodeXID::is_xid_start(*c) => true,
        '$' | '_' => true,
        _ => false,
    })(s)
}

pub fn identifier_continue(s: Span) -> ParseResult<char> {
    verify(alt((identifier_unicode_esc_sequence, anychar)), |c: &char| match c {
        // These are already matched by is_xid_continue but are explicitly at the beginning for performance reasons.
        'a'..='z' => true,
        'A'..='Z' => true,
        c if unicode_xid::UnicodeXID::is_xid_continue(*c) => true,
        '$' | '_' => true,
        _ => false,
    })(s)
}

/// ```ebnf
/// IdentifierName
/// ```
pub fn parse_identifier_name(s: Span) -> ParseResult<Node> {
    context("identifier", |s| {let (s, start) = position(s)?;
    let (s, name) =
        map(
            pair(identifier_start, many0(identifier_continue)),
            |(c, v)| format!("{}{}", c, v.into_iter().collect::<String>()),
        )(s)?;
    let (s, end) = position(s)?;

    let (s, _) = sp0(s)?;
    Ok((s, NodeKind::Identifier { name }.with_pos(start, end)))})(s)
}

/// ```ebnf
/// Identifier :: IdentifierName but not ReservedWord
/// ```
pub fn parse_identifier(s: Span) -> ParseResult<Node> {
    context("identifier", |s| {let (s, start) = position(s)?;
    let (s, name) = verify(
        map(
            pair(identifier_start, many0(identifier_continue)),
            |(c, v)| format!("{}{}", c, v.into_iter().collect::<String>()),
        ),
        |ident: &str| 
            // make sure identifier is not a reserved word
            // not(reserved_word2)(ident.into()).is_ok(),
            not(reserved_word)(ident.into()).is_ok()
    )(s)?;
    let (s, end) = position(s)?;

    let (s, _) = sp0(s)?;
    Ok((s, NodeKind::Identifier { name }.with_pos(start, end)))})(s)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_identifier() {
        all_consuming(parse_identifier)("myVar".into()).unwrap();
        all_consuming(parse_identifier)("abc123".into()).unwrap();
        all_consuming(parse_identifier)("$".into()).unwrap();
        all_consuming(parse_identifier)("_".into()).unwrap();
        all_consuming(parse_identifier)("$elem".into()).unwrap();
        all_consuming(parse_identifier)("_elem".into()).unwrap();
        all_consuming(parse_identifier)("$123".into()).unwrap();
        all_consuming(parse_identifier)("_123".into()).unwrap();

        // identifiers can start with keywords
        all_consuming(parse_identifier)("var_id".into()).unwrap();
        all_consuming(parse_identifier)("class_id".into()).unwrap();
    }

    #[test]
    fn test_identifier_with_trailing_whitespace() {
        all_consuming(parse_identifier)("abc123 ".into()).unwrap();
        all_consuming(parse_identifier)("abc123 \t\n ".into()).unwrap();
    }

    #[test]
    fn test_bad_identifier() {
        all_consuming(parse_identifier)("123abc".into()).unwrap_err();
    }

    #[test]
    fn test_keyword_not_identifier() {
        all_consuming(parse_identifier)("var".into()).unwrap_err();
        all_consuming(parse_identifier)("class".into()).unwrap_err();
        all_consuming(parse_identifier)("true".into()).unwrap_err();
        all_consuming(parse_identifier)("null".into()).unwrap_err();
    }

    #[test]
    fn test_identifier_with_unicode_escape() {
        assert_eq!(
            parse_identifier(r#"abc\u0061"#.into()).unwrap().1.kind,
            NodeKind::Identifier {
                name: "abca".into()
            }
        );
        assert_eq!(
            parse_identifier(r#"abc\u0061\u0062123"#.into())
                .unwrap()
                .1
                .kind,
            NodeKind::Identifier {
                name: "abcab123".into()
            }
        );
    }
}
