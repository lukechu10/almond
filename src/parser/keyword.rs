//! Parsing for JS keywords

use crate::parser::util::*;
use crate::parser::*;

/// Succeeds if parsed a reserved word. Should be used with `not` to check if an identifier is not a reserved word.
pub fn parse_reserved_word(s: Span) -> ParseResult<()> {
    alt((
        parse_keyword,
        parse_future_reserved_word,
        value((), ws0(null_lit)),
        value((), ws0(bool_lit)),
    ))(s)
}

/// Succeeds if parsed is a keyword. Use `reserved_word` with `not` instead to check if an identifier is not a reserved word.
pub fn parse_keyword(s: Span) -> ParseResult<()> {
    ws0(alt((
        keyword_break,
        keyword_do,
        keyword_instanceof,
        keyword_typeof,
        keyword_case,
        keyword_else,
        keyword_new,
        keyword_var,
        keyword_catch,
        keyword_finally,
        keyword_return,
        keyword_void,
        keyword_continue,
        keyword_for,
        keyword_switch,
        keyword_while,
        keyword_debugger,
        keyword_function,
        keyword_this,
        keyword_with,
        alt((
            keyword_default,
            keyword_if,
            keyword_throw,
            keyword_delete,
            keyword_in,
            keyword_try,
        )),
    )))(s)
}

pub fn parse_future_reserved_word_lax(s: Span) -> ParseResult<()> {
    ws0(alt((
        keyword_class,
        keyword_enum,
        keyword_extends,
        keyword_super,
        keyword_const,
        keyword_export,
        keyword_import,
        keyword_await,
    )))(s)
}

pub fn parse_future_reserved_word_strict(s: Span) -> ParseResult<()> {
    ws0(alt((
        parse_future_reserved_word_lax,
        keyword_implements,
        keyword_let,
        keyword_private,
        keyword_public,
        keyword_interface,
        keyword_package,
        keyword_protected,
        keyword_static,
        keyword_yield,
    )))(s)
}

pub fn parse_future_reserved_word(s: Span) -> ParseResult<()> {
    parse_future_reserved_word_strict(s)
}

pub fn keyword_break(s: Span) -> ParseResult<()> {
    value((), pair(tag("break"), not(identifier_continue)))(s)
}
pub fn keyword_do(s: Span) -> ParseResult<()> {
    value((), pair(tag("do"), not(identifier_continue)))(s)
}
pub fn keyword_instanceof(s: Span) -> ParseResult<()> {
    value((), pair(tag("instanceof"), not(identifier_continue)))(s)
}
pub fn keyword_typeof(s: Span) -> ParseResult<()> {
    value((), pair(tag("typeof"), not(identifier_continue)))(s)
}
pub fn keyword_case(s: Span) -> ParseResult<()> {
    value((), pair(tag("case"), not(identifier_continue)))(s)
}
pub fn keyword_else(s: Span) -> ParseResult<()> {
    value((), pair(tag("else"), not(identifier_continue)))(s)
}
pub fn keyword_new(s: Span) -> ParseResult<()> {
    value((), pair(tag("new"), not(identifier_continue)))(s)
}
pub fn keyword_var(s: Span) -> ParseResult<()> {
    value((), pair(tag("var"), not(identifier_continue)))(s)
}
pub fn keyword_catch(s: Span) -> ParseResult<()> {
    value((), pair(tag("catch"), not(identifier_continue)))(s)
}
pub fn keyword_finally(s: Span) -> ParseResult<()> {
    value((), pair(tag("finally"), not(identifier_continue)))(s)
}
pub fn keyword_return(s: Span) -> ParseResult<()> {
    value((), pair(tag("return"), not(identifier_continue)))(s)
}
pub fn keyword_void(s: Span) -> ParseResult<()> {
    value((), pair(tag("void"), not(identifier_continue)))(s)
}
pub fn keyword_continue(s: Span) -> ParseResult<()> {
    value((), pair(tag("continue"), not(identifier_continue)))(s)
}
pub fn keyword_for(s: Span) -> ParseResult<()> {
    value((), pair(tag("for"), not(identifier_continue)))(s)
}
pub fn keyword_switch(s: Span) -> ParseResult<()> {
    value((), pair(tag("switch"), not(identifier_continue)))(s)
}
pub fn keyword_while(s: Span) -> ParseResult<()> {
    value((), pair(tag("while"), not(identifier_continue)))(s)
}
pub fn keyword_debugger(s: Span) -> ParseResult<()> {
    value((), pair(tag("debugger"), not(identifier_continue)))(s)
}
pub fn keyword_function(s: Span) -> ParseResult<()> {
    value((), pair(tag("function"), not(identifier_continue)))(s)
}
pub fn keyword_this(s: Span) -> ParseResult<()> {
    value((), pair(tag("this"), not(identifier_continue)))(s)
}
pub fn keyword_with(s: Span) -> ParseResult<()> {
    value((), pair(tag("with"), not(identifier_continue)))(s)
}
pub fn keyword_default(s: Span) -> ParseResult<()> {
    value((), pair(tag("default"), not(identifier_continue)))(s)
}
pub fn keyword_if(s: Span) -> ParseResult<()> {
    value((), pair(tag("if"), not(identifier_continue)))(s)
}
pub fn keyword_throw(s: Span) -> ParseResult<()> {
    value((), pair(tag("throw"), not(identifier_continue)))(s)
}
pub fn keyword_delete(s: Span) -> ParseResult<()> {
    value((), pair(tag("delete"), not(identifier_continue)))(s)
}
pub fn keyword_in(s: Span) -> ParseResult<()> {
    value((), pair(tag("in"), not(identifier_continue)))(s)
}
pub fn keyword_try(s: Span) -> ParseResult<()> {
    value((), pair(tag("try"), not(identifier_continue)))(s)
}
pub fn keyword_get(s: Span) -> ParseResult<()> {
    value((), pair(tag("get"), not(identifier_continue)))(s)
}
pub fn keyword_set(s: Span) -> ParseResult<()> {
    value((), pair(tag("set"), not(identifier_continue)))(s)
}
pub fn keyword_class(s: Span) -> ParseResult<()> {
    value((), pair(tag("class"), not(identifier_continue)))(s)
}
pub fn keyword_enum(s: Span) -> ParseResult<()> {
    value((), pair(tag("enum"), not(identifier_continue)))(s)
}
pub fn keyword_extends(s: Span) -> ParseResult<()> {
    value((), pair(tag("extends"), not(identifier_continue)))(s)
}
pub fn keyword_super(s: Span) -> ParseResult<()> {
    value((), pair(tag("super"), not(identifier_continue)))(s)
}
pub fn keyword_const(s: Span) -> ParseResult<()> {
    value((), pair(tag("const"), not(identifier_continue)))(s)
}
pub fn keyword_export(s: Span) -> ParseResult<()> {
    value((), pair(tag("export"), not(identifier_continue)))(s)
}
pub fn keyword_import(s: Span) -> ParseResult<()> {
    value((), pair(tag("import"), not(identifier_continue)))(s)
}
pub fn keyword_implements(s: Span) -> ParseResult<()> {
    value((), pair(tag("implements"), not(identifier_continue)))(s)
}
pub fn keyword_let(s: Span) -> ParseResult<()> {
    value((), pair(tag("let"), not(identifier_continue)))(s)
}
pub fn keyword_private(s: Span) -> ParseResult<()> {
    value((), pair(tag("private"), not(identifier_continue)))(s)
}
pub fn keyword_public(s: Span) -> ParseResult<()> {
    value((), pair(tag("public"), not(identifier_continue)))(s)
}
pub fn keyword_interface(s: Span) -> ParseResult<()> {
    value((), pair(tag("interface"), not(identifier_continue)))(s)
}
pub fn keyword_package(s: Span) -> ParseResult<()> {
    value((), pair(tag("package"), not(identifier_continue)))(s)
}
pub fn keyword_protected(s: Span) -> ParseResult<()> {
    value((), pair(tag("protected"), not(identifier_continue)))(s)
}
pub fn keyword_static(s: Span) -> ParseResult<()> {
    value((), pair(tag("static"), not(identifier_continue)))(s)
}
pub fn keyword_yield(s: Span) -> ParseResult<()> {
    value((), pair(tag("yield"), not(identifier_continue)))(s)
}
// es2017
pub fn keyword_async(s: Span) -> ParseResult<()> {
    value((), pair(tag("async"), not(identifier_continue)))(s)
}
pub fn keyword_await(s: Span) -> ParseResult<()> {
    value((), pair(tag("await"), not(identifier_continue)))(s)
}
