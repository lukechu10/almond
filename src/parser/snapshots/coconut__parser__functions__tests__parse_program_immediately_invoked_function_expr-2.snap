---
source: src/parser/functions.rs
expression: "parse_program(r#\"(function (global, factory) {\n                    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :\n                        typeof define === 'function' && define.amd ? define(['exports'], factory) :\n                            (global = global || self, factory(global.Library = {}));\n                }(this, (function (exports) {})));\"#.into()).unwrap().1"
---
{
  "type": "Program",
  "body": [
    {
      "type": "ExpressionStatement",
      "expression": {
        "type": "CallExpression",
        "callee": {
          "type": "FunctionExpression",
          "id": null,
          "params": [
            {
              "type": "Identifier",
              "name": "global",
              "start": 11,
              "end": 17
            },
            {
              "type": "Identifier",
              "name": "factory",
              "start": 19,
              "end": 26
            }
          ],
          "body": {
            "type": "BlockStatement",
            "body": [
              {
                "type": "ExpressionStatement",
                "expression": {
                  "type": "ConditionalExpression",
                  "test": {
                    "type": "LogicalExpression",
                    "operator": "&&",
                    "left": {
                      "type": "BinaryExpression",
                      "operator": "===",
                      "left": {
                        "type": "UnaryExpression",
                        "operator": "typeof",
                        "prefix": true,
                        "argument": {
                          "type": "Identifier",
                          "name": "exports",
                          "start": 57,
                          "end": 64
                        },
                        "start": 50,
                        "end": 65
                      },
                      "right": {
                        "type": "Literal",
                        "value": "object",
                        "start": 69,
                        "end": 77
                      },
                      "start": 50,
                      "end": 77
                    },
                    "right": {
                      "type": "BinaryExpression",
                      "operator": "!==",
                      "left": {
                        "type": "UnaryExpression",
                        "operator": "typeof",
                        "prefix": true,
                        "argument": {
                          "type": "Identifier",
                          "name": "module",
                          "start": 88,
                          "end": 94
                        },
                        "start": 81,
                        "end": 95
                      },
                      "right": {
                        "type": "Literal",
                        "value": "undefined",
                        "start": 99,
                        "end": 110
                      },
                      "start": 81,
                      "end": 110
                    },
                    "start": 50,
                    "end": 110
                  },
                  "consequent": {
                    "type": "CallExpression",
                    "callee": {
                      "type": "Identifier",
                      "name": "factory",
                      "start": 113,
                      "end": 120
                    },
                    "arguments": [
                      {
                        "type": "Identifier",
                        "name": "exports",
                        "start": 121,
                        "end": 128
                      }
                    ],
                    "start": 113,
                    "end": 129
                  },
                  "alternate": {
                    "type": "ConditionalExpression",
                    "test": {
                      "type": "LogicalExpression",
                      "operator": "&&",
                      "left": {
                        "type": "BinaryExpression",
                        "operator": "===",
                        "left": {
                          "type": "UnaryExpression",
                          "operator": "typeof",
                          "prefix": true,
                          "argument": {
                            "type": "Identifier",
                            "name": "define",
                            "start": 163,
                            "end": 169
                          },
                          "start": 156,
                          "end": 170
                        },
                        "right": {
                          "type": "Literal",
                          "value": "function",
                          "start": 174,
                          "end": 184
                        },
                        "start": 156,
                        "end": 184
                      },
                      "right": {
                        "type": "MemberExpression",
                        "object": {
                          "type": "Identifier",
                          "name": "define",
                          "start": 188,
                          "end": 194
                        },
                        "property": {
                          "type": "Identifier",
                          "name": "amd",
                          "start": 195,
                          "end": 198
                        },
                        "computed": false,
                        "start": 188,
                        "end": 198
                      },
                      "start": 156,
                      "end": 198
                    },
                    "consequent": {
                      "type": "CallExpression",
                      "callee": {
                        "type": "Identifier",
                        "name": "define",
                        "start": 201,
                        "end": 207
                      },
                      "arguments": [
                        {
                          "type": "ArrayExpression",
                          "elements": [
                            {
                              "type": "Literal",
                              "value": "exports",
                              "start": 209,
                              "end": 218
                            }
                          ],
                          "start": 208,
                          "end": 219
                        },
                        {
                          "type": "Identifier",
                          "name": "factory",
                          "start": 221,
                          "end": 228
                        }
                      ],
                      "start": 201,
                      "end": 229
                    },
                    "alternate": {
                      "type": "SequenceExpression",
                      "expressions": [
                        {
                          "type": "AssignmentExpression",
                          "operator": "=",
                          "left": {
                            "type": "Identifier",
                            "name": "global",
                            "start": 261,
                            "end": 267
                          },
                          "right": {
                            "type": "LogicalExpression",
                            "operator": "||",
                            "left": {
                              "type": "Identifier",
                              "name": "global",
                              "start": 270,
                              "end": 276
                            },
                            "right": {
                              "type": "Identifier",
                              "name": "self",
                              "start": 280,
                              "end": 284
                            },
                            "start": 270,
                            "end": 284
                          },
                          "start": 261,
                          "end": 284
                        },
                        {
                          "type": "CallExpression",
                          "callee": {
                            "type": "Identifier",
                            "name": "factory",
                            "start": 286,
                            "end": 293
                          },
                          "arguments": [
                            {
                              "type": "AssignmentExpression",
                              "operator": "=",
                              "left": {
                                "type": "MemberExpression",
                                "object": {
                                  "type": "Identifier",
                                  "name": "global",
                                  "start": 294,
                                  "end": 300
                                },
                                "property": {
                                  "type": "Identifier",
                                  "name": "Library",
                                  "start": 301,
                                  "end": 308
                                },
                                "computed": false,
                                "start": 294,
                                "end": 308
                              },
                              "right": {
                                "type": "ObjectExpression",
                                "properties": [],
                                "start": 311,
                                "end": 313
                              },
                              "start": 294,
                              "end": 313
                            }
                          ],
                          "start": 286,
                          "end": 314
                        }
                      ],
                      "start": 261,
                      "end": 314
                    },
                    "start": 156,
                    "end": 314
                  },
                  "start": 50,
                  "end": 314
                },
                "start": 50,
                "end": 316
              }
            ],
            "start": 28,
            "end": 334
          },
          "async": false,
          "start": 1,
          "end": 334
        },
        "arguments": [
          {
            "type": "ThisExpression",
            "start": 335,
            "end": 339
          },
          {
            "type": "FunctionExpression",
            "id": null,
            "params": [
              {
                "type": "Identifier",
                "name": "exports",
                "start": 352,
                "end": 359
              }
            ],
            "body": {
              "type": "BlockStatement",
              "body": [],
              "start": 361,
              "end": 363
            },
            "async": false,
            "start": 342,
            "end": 363
          }
        ],
        "start": 1,
        "end": 365
      },
      "start": 0,
      "end": 367
    }
  ],
  "start": 0,
  "end": 367
}
